src/core/batch.rkt

```diff
@@ -82,71 +82,173 @@
-        [(list 'pow base 2/3) `(cbrt ,(munge `(* ,base ,base)))]
-        [(list 'pow base power)
-         #:when (exact-integer? power)
-         `(pow ,(munge base) ,(munge power))]
-        [(list 'pow base power) `(exp ,(munge `(* ,power (log ,base))))]
-        [(list 'tan args) `(/ ,(munge `(sin ,args)) ,(munge `(cos ,args)))]
-        [(list 'cosh args) `(* ,(munge 1/2) ,(munge `(+ (exp ,args) (/ 1 (exp ,args)))))]
-        [(list 'sinh args) `(* ,(munge 1/2) ,(munge `(+ (exp ,args) (neg (/ 1 (exp ,args))))))]
-        [(list 'tanh args)
-         `(/ ,(munge `(+ (exp ,args) (neg (/ 1 (exp ,args)))))
-             ,(munge `(+ (exp ,args) (/ 1 (exp ,args)))))]
-        [(list 'asinh args) `(log ,(munge `(+ ,args (sqrt (+ (* ,args ,args) 1)))))]
-        [(list 'acosh args) `(log ,(munge `(+ ,args (sqrt (+ (* ,args ,args) -1)))))]
-        [(list 'atanh args) `(* ,(munge 1/2) ,(munge `(log (/ (+ 1 ,args) (+ 1 (neg ,args))))))]
-        [(list op args ...) (cons op (map munge args))]
-        [(approx spec impl) (approx spec (munge impl))]
-        [_ prog]))
+  ; Mapping from nodes to nodes*
+  (define mappings
+    (make-hash (map (λ (n) (cons n n)) (build-list (batch-nodes-length input-batch) values))))
+
+  ; Adding a node to hash
+  (define (append-node node)
     (hash-ref! exprhash
                node
                (lambda ()
                  (begin0 (+ exprc varc) ; store in cache, update exprs, exprc
                    (set! exprc (+ 1 exprc))
                    (set! icache (cons node icache))))))
 
-  (set-batch-roots! batch (list->vector (map munge (batch->progs batch))))
-  (set-batch-nodes! batch (list->vector (reverse icache)))
-  (set-batch-nodes-length! batch (vector-length (batch-nodes batch))))
+  ; Sequential rewriting
+  (for ([node (in-vector nodes)]
+        [n (in-naturals)])
+    (match node
+      [(list '- arg1 arg2)
+       (define neg-index (append-node `(neg ,(hash-ref mappings arg2))))
+       (hash-set! mappings n (append-node `(+ ,(hash-ref mappings arg1) ,neg-index)))]
+      [(list 'pow base power)
+       #:when (equal? (vector-ref nodes power) 1/2) ; 1/2 is to be removed from exprhash
+       (hash-set! mappings n (append-node `(sqrt ,(hash-ref mappings base))))]
+      [(list 'pow base power)
+       #:when (equal? (vector-ref nodes power) 1/3) ; 1/3 is to be removed from exprhash
+       (hash-set! mappings n (append-node `(cbrt ,(hash-ref mappings base))))]
+      [(list 'pow base power)
+       #:when (equal? (vector-ref nodes power) 2/3) ; 2/3 is to be removed from exprhash
+       (define mult-index (append-node `(* ,(hash-ref mappings base) ,(hash-ref mappings base))))
+       (hash-set! mappings n (append-node `(cbrt ,mult-index)))]
+      [(list 'pow base power)
+       #:when (exact-integer? (vector-ref nodes power))
+       (hash-set! mappings
+                  n
+                  (append-node `(pow ,(hash-ref mappings base) ,(hash-ref mappings power))))]
```

Comment:
We could be a bit more clever here, like append a `(pown ,(vector-ref nodes power) ,base)` and then avoid a lookup during Taylor. This is not a big deal, just a thought I had while reading.

Reply:
Where "pown" is what IEEE 754 calls the integer power function.

Reply:
Make sense, less work at `taylor`. Changed to `pown`